sdk/python/kfp/components/structures/components.json_schema.json

```diff
@@ -0,0 +1,402 @@
+      "properties": {
+        "inputValue" : {
+          "description": "Name of the input.",
+          "type": "string"
+        }
+      },
+      "additionalProperties": false
+    },
+
+    "InputPathPlaceholder": {
+      "description": "Represents the command-line argument placeholder that will be replaced at run-time by a local file path pointing to a file containing the input argument value.",
+      "type": "object",
+      "required": ["inputPath"],
+      "properties": {
+        "inputPath" : {
+          "description": "Name of the input.",
+          "type": "string"
+        }
+      },
+      "additionalProperties": false
+    },
+
+    "OutputPathPlaceholder": {
+      "description": "Represents the command-line argument placeholder that will be replaced at run-time by a local file path pointing to a file where the program should write its output data.",
+      "type": "object",
+      "required": ["outputPath"],
+      "properties": {
+        "outputPath" : {
+          "description": "Name of the output.",
+          "type": "string"
+        }
+      },
+      "additionalProperties": false
+    },
+
+    "CommandlineArgumentType": {
+      "oneOf": [
+        {"$ref": "#/definitions/PrimitiveTypes"},
+        {"$ref": "#/definitions/InputValuePlaceholder"},
+        {"$ref": "#/definitions/InputPathPlaceholder"},
+        {"$ref": "#/definitions/OutputPathPlaceholder"},
+        {"$ref": "#/definitions/ConcatPlaceholder"},
+        {"$ref": "#/definitions/IfPlaceholder"}
+      ]
+    },
+
+    "CommandlineArgumentOrArrayType": {
+      "oneOf": [
+        {"$ref": "#/definitions/CommandlineArgumentType"},
+        {"type": "array", "items": {"$ref": "#/definitions/CommandlineArgumentType"}}
+      ]
+    },
+
+    "ConcatPlaceholder": {
+      "description": "Represents the command-line argument placeholder that will be replaced at run-time by the concatenated values of its items.",
+      "type": "object",
+      "required": ["concat"],
+      "properties": {
+        "concat" : {
+          "description": "Items to concatenate",
+          "type": "array",
+          "items": {"$ref": "#/definitions/CommandlineArgumentType"}
+        }
+      },
+      "additionalProperties": false
+    },
+
+    "IsPresentPlaceholder": {
+      "description": "Represents the command-line argument placeholder that will be replaced at run-time by a boolean value specifying whether the caller has passed an argument for the specified optional input.",
+      "type": "object",
+      "properties": {
+        "isPresent": {
+          "description": "Name of the input.",
+          "type": "string"
+        }
+      },
+      "additionalProperties": false
+    },
+
+    "IfConditionArgumentType": {
+      "oneOf": [
+        {"$ref": "#/definitions/IsPresentPlaceholder"},
+        {"type": "boolean"},
+        {"type": "string"},
+        {"$ref": "#/definitions/InputValuePlaceholder"}
+      ]
+    },
+
+    "IfPlaceholder": {
```

Comment:
These placeholders are almost implementing a templating language within this schema. Do we need this complexity from the get-go, or is it possible to start simple, and add such functionality when needed? Also, to echo @jlewi's concern, have we considered if we can get such templating capabilities from something like Jsonnet?

Reply:
If we go with JSON/YAML-based placeholders, the spec is future-proof and reliable. Going with JSON/YAML structural placeholders allows us for instance to support Jsonnet or any other templating system in future.
Going with text templating will close that road - we'll be stuck supporting the complicated and broken text templating system forever; Any new feature would risk breaking existing components and users.

P.S. I can remove `If` from the initial spec to reduce the schema complexity, but it's pretty strange to remove working and tested code, because you could use it for something complex.